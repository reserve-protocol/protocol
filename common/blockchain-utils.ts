import { BigNumber } from 'ethers'
import { HardhatRuntimeEnvironment } from 'hardhat/types'

// getChainId: Returns current chain Id
export const getChainId = async (hre: HardhatRuntimeEnvironment) => {
  let _chainId
  try {
    _chainId = await hre.network.provider.send('eth_chainId')
  } catch (e) {
    console.log('failed to get chainId, falling back on net_version...')
    _chainId = await hre.network.provider.send('net_version')
  }

  if (!_chainId) {
    throw new Error(`could not get chainId from network`)
  }
  if (_chainId.startsWith('0x')) {
    _chainId = BigNumber.from(_chainId).toString()
  }
  return _chainId
}
