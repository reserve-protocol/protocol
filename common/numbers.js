const { BigNumber } = require("ethers");
const { BN_SCALE_FACTOR } = require("./constants");

const bn = (x) => {
    if (BigNumber.isBigNumber(x)) return x;
    const stringified = parseScientific(x.toString());
    const integer = stringified.split('.')[0];
    return BigNumber.from(integer);
};

const fp = (x) => {
    return BN_SCALE_FACTOR.mul(x);
}
 
function parseScientific(num) {
    // If the number is not in scientific notation return it as it is
    if (!/\d+\.?\d*e[+-]*\d+/i.test(num)) return num;

    // Remove the sign
    const numberSign = Math.sign(Number(num));
    num = Math.abs(Number(num)).toString();

    // Parse into coefficient and exponent
    const [coefficient, exponent] = num.toLowerCase().split('e');
    let zeros = Math.abs(Number(exponent));
    const exponentSign = Math.sign(Number(exponent));
    const [integer, decimals] = (coefficient.indexOf('.') != -1 ? coefficient : `${coefficient}.`).split('.');

    if (exponentSign === -1) {
        zeros -= integer.length;
        num =
            zeros < 0
                ? integer.slice(0, zeros) + '.' + integer.slice(zeros) + decimals
                : '0.' + '0'.repeat(zeros) + integer + decimals;
    } else {
        if (decimals) zeros -= decimals.length;
        num =
            zeros < 0
                ? integer + decimals.slice(0, zeros) + '.' + decimals.slice(zeros)
                : integer + decimals + '0'.repeat(zeros);
    }

    return numberSign < 0 ? '-' + num : num;
}

module.exports = {
    bn,
    fp
}