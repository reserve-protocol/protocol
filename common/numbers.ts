import { BigNumber } from 'ethers'
import { ethers } from 'hardhat'
import { BN_SCALE_FACTOR, SCALE_DECIMALS } from './constants'

export const ZERO = BigNumber.from(0)

export const bn = (x: BigNumber | number): BigNumber => {
  if (BigNumber.isBigNumber(x)) return x
  const stringified = parseScientific(x.toString())
  const integer = stringified.split('.')[0]
  return BigNumber.from(integer)
}

export const pow10 = (exponent: number): BigNumber => {
  return bn(10).pow(exponent)
}

export const fp = (x: BigNumber | number): BigNumber => {
  if (BigNumber.isBigNumber(x)) {
    return BN_SCALE_FACTOR.mul(x)
  }
  return ethers.utils.parseUnits(x.toString(), SCALE_DECIMALS)
}

export const divCeil = (x: BigNumber, y: BigNumber): BigNumber =>
  // ceil(x/y) == (x + y - 1) / y
  x.add(y).sub(1).div(y)

function parseScientific(num: string) {
  // If the number is not in scientific notation return it as it is
  if (!/\d+\.?\d*e[+-]*\d+/i.test(num)) return num

  // Remove the sign
  const numberSign = Math.sign(Number(num))
  num = Math.abs(Number(num)).toString()

  // Parse into coefficient and exponent
  const [coefficient, exponent] = num.toLowerCase().split('e')
  let zeros = Math.abs(Number(exponent))
  const exponentSign = Math.sign(Number(exponent))
  const [integer, decimals] = (coefficient.indexOf('.') != -1 ? coefficient : `${coefficient}.`).split('.')

  if (exponentSign === -1) {
    zeros -= integer.length
    num =
      zeros < 0
        ? integer.slice(0, zeros) + '.' + integer.slice(zeros) + decimals
        : '0.' + '0'.repeat(zeros) + integer + decimals
  } else {
    if (decimals) zeros -= decimals.length
    num =
      zeros < 0
        ? integer + decimals.slice(0, zeros) + '.' + decimals.slice(zeros)
        : integer + decimals + '0'.repeat(zeros)
  }

  return numberSign < 0 ? '-' + num : num
}
