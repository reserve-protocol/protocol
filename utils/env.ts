import dotenv from 'dotenv'

dotenv.config()

type IEnvVars =
  | 'MAINNET_RPC_URL'
  | 'ALCHEMY_MAINNET_RPC_URL'
  | 'GOERLI_RPC_URL'
  | 'MNEMONIC'
  | 'REPORT_GAS'
  | 'FORK'
  | 'SLOW'
  | 'PROTO'
  | 'PROTO_IMPL'
  | 'ETHERSCAN_API_KEY'
  | 'NO_OPT'
  | 'MAINNET_BLOCK'
  | 'ONLY_FAST'
  | 'JOBS'

export function useEnv(key: IEnvVars | IEnvVars[], _default = ''): string {
  if (typeof key === 'string') {
    return process.env[key] ?? _default
  }
  if (typeof key === 'object') {
    for (const s of key) {
      if (process.env[s]) {
        return process.env[s]!
      }
    }
  }
  return _default
}
