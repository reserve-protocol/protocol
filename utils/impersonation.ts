import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers'
import { HardhatRuntimeEnvironment } from 'hardhat/types'

type ImpersonationFunction<T> = (signer: SignerWithAddress) => Promise<T>

/* whileImpersonating(address, f):

   Set up `signer` to be an ethers transaction signer that impersonates the account address
   `address`. In that context, call f(signer). `address` can be either a contract address or an
   external account, so you can use often this instead of building entire mock contracts.

   Example usage:

   await whileImpersonating(basketHandler.address, async (signer) => {
     await expect(rToken.connect(signer).setBasketsNeeded(fp('1'))
     .to.emit(rToken, 'BasketsNeededChanged')
   })

   This does the following:
   - Sets the basketHandler Eth balance to 2^256-1 (so it has plenty of gas)
   - Calls rToken.setBasketsNeeded _as_ the basketHandler contract,
   - Checks that that call emits the event 'BasketNeededChanged'
*/
export const whileImpersonating = async (
  hre: HardhatRuntimeEnvironment,
  address: string,
  f: ImpersonationFunction<void>
) => {
  // Set maximum ether balance at address
  await hre.network.provider.request({
    method: 'hardhat_setBalance',
    params: [address, '0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff'],
  })
  await hre.network.provider.request({
    method: 'hardhat_impersonateAccount',
    params: [address],
  })
  const signer = await hre.ethers.getSigner(address)

  await f(signer)

  await hre.network.provider.request({
    method: 'hardhat_stopImpersonatingAccount',
    params: [address],
  })
  // If anyone ever needs it, we could make sure here that we set the balance at address back to
  // its original quantity...
}
