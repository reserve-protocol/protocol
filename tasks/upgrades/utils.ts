import { HardhatRuntimeEnvironment } from 'hardhat/types'
import { IComponents } from '../../common/configuration'
import { IDeployments } from '../../scripts/deployment/common'
import { isValidContract } from '../../common/blockchain-utils'

export const validateDeployments = async (
  hre: HardhatRuntimeEnvironment,
  deployments: IDeployments,
  version: string
) => {
  // Check trading lib defined
  if (!deployments.tradingLib) {
    throw new Error(
      `Missing deployed TradingLib for version ${version} in network ${hre.network.name}`
    )
  } else if (!(await isValidContract(hre, deployments.tradingLib))) {
    throw new Error(`TradingLib contract not found in network ${hre.network.name}`)
  }

  // Check Main implementation is defined
  if (!deployments.implementations.main) {
    throw new Error(
      `Missing deployed Main implementation for version ${version} in network ${hre.network.name}`
    )
  } else if (!(await isValidContract(hre, deployments.implementations.main))) {
    throw new Error(`Main contract not found in network ${hre.network.name}`)
  }

  // Check all componet implementations are defined
  const componentsKeys: string[] = [
    'backingManager',
    'basketHandler',
    'broker',
    'distributor',
    'furnace',
    'rTokenTrader',
    'rsrTrader',
    'rToken',
    'stRSR',
  ]

  componentsKeys.forEach(async (keystr) => {
    if (!deployments.implementations.components[keystr as keyof IComponents]) {
      throw new Error(
        `Missing deployed ${keystr} implementation for version ${version} in network ${hre.network.name}`
      )
    } else if (
      !(await isValidContract(
        hre,
        deployments.implementations.components[keystr as keyof IComponents]
      ))
    ) {
      throw new Error(
        `Implementation ${keystr} contract not found for version ${version} in network ${hre.network.name}`
      )
    }
  })
}
