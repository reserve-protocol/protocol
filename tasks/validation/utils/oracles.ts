/* eslint-disable no-empty */
import { networkConfig } from '../../../common/configuration'
import { EACAggregatorProxyMock } from '@typechain/EACAggregatorProxyMock'
import { HardhatRuntimeEnvironment } from 'hardhat/types'
import { BigNumber } from 'ethers'
import { AggregatorV3Interface } from '@typechain/index'
import { ONE_ADDRESS } from '../../../common/constants'

export const overrideOracle = async (
  hre: HardhatRuntimeEnvironment,
  oracleAddress: string
): Promise<EACAggregatorProxyMock> => {
  const oracle = await hre.ethers.getContractAt(
    'contracts/plugins/mocks/EACAggregatorProxyMock.sol:EACAggregatorProxy',
    oracleAddress
  )
  const aggregator = await oracle.aggregator()
  const accessController = await oracle.accessController()
  const initPrice = await oracle.latestRoundData()
  const mockOracleFactory = await hre.ethers.getContractFactory('EACAggregatorProxyMock')
  const mockOracle = await mockOracleFactory.deploy(aggregator, accessController, initPrice.answer)
  const bytecode = await hre.network.provider.send('eth_getCode', [mockOracle.address, 'latest'])
  await hre.network.provider.request({
    method: 'hardhat_setCode',
    params: [oracleAddress, bytecode],
  })
  return hre.ethers.getContractAt('EACAggregatorProxyMock', oracleAddress)
}

export const pushOraclesForward = async (
  hre: HardhatRuntimeEnvironment,
  rTokenAddress: string,
  extraAssets: string[] = []
) => {
  console.log(`🔃 Pushing Oracles forward for RToken: ${rTokenAddress}`)
  const rToken = await hre.ethers.getContractAt('RTokenP1', rTokenAddress)
  const main = await hre.ethers.getContractAt('IMain', await rToken.main())
  const assetRegistry = await hre.ethers.getContractAt(
    'AssetRegistryP1',
    await main.assetRegistry()
  )
  const registry = await assetRegistry.getRegistry()

  let addresses: string[] = [] // hacky way to ensure only unique updates to save time
  for (const asset of registry.assets) {
    addresses = await pushOracleForward(hre, asset, addresses)
  }

  for (const asset of extraAssets) {
    await pushOracleForward(hre, asset, addresses)
  }
}

export const pushOracleForward = async (
  hre: HardhatRuntimeEnvironment,
  asset: string,
  addresses: string[]
): Promise<string[]> => {
  // Need to handle all oracle cases, ie targetUnitChainlinkFeed, PoolTokens, etc
  const updateAnswer = async (chainlinkFeed: AggregatorV3Interface) => {
    if (addresses.indexOf(chainlinkFeed.address) != -1) return
    const initPrice = await chainlinkFeed.latestRoundData()
    const oracle = await overrideOracle(hre, chainlinkFeed.address)
    await oracle.updateAnswer(initPrice.answer)

    addresses.push(chainlinkFeed.address)
    console.log('✅ Feed Updated:', chainlinkFeed.address)
  }

  // chainlinkFeed
  try {
    const assetContract = await hre.ethers.getContractAt('TestIAsset', asset)
    const feed = await hre.ethers.getContractAt(
      'AggregatorV3Interface',
      await assetContract.chainlinkFeed()
    )
    if (feed.address != ONE_ADDRESS) await updateAnswer(feed)
  } catch {
    // console.error('❌ chainlinkFeed not found for:', asset, 'skipping...')
  }

  // targetUnitChainlinkFeed
  try {
    const assetContractNonFiat = await hre.ethers.getContractAt('NonFiatCollateral', asset)
    const feed = await hre.ethers.getContractAt(
      'AggregatorV3Interface',
      await assetContractNonFiat.targetUnitChainlinkFeed()
    )
    await updateAnswer(feed)
  } catch {
    // console.error('❌ targetUnitChainlinkFeed not found for:', asset, 'skipping...')
  }

  // targetPerRefChainlinkFeed, uoaPerTargetChainlinkFeed, refPerTokenChainlinkFeed
  try {
    const assetContractLido = await hre.ethers.getContractAt('L2LSDCollateral', asset)
    const feed = await hre.ethers.getContractAt(
      'AggregatorV3Interface',
      await assetContractLido.exchangeRateChainlinkFeed()
    )
    await updateAnswer(feed)
  } catch {
    // console.error('❌ exchangeRateChainlinkFeed not found for:', asset, 'skipping...')
  }

  // targetPerRefChainlinkFeed, uoaPerTargetChainlinkFeed, refPerTokenChainlinkFeed
  try {
    const assetContractLido = await hre.ethers.getContractAt('L2LidoStakedEthCollateral', asset)
    let feed = await hre.ethers.getContractAt(
      'AggregatorV3Interface',
      await assetContractLido.targetPerRefChainlinkFeed()
    )
    await updateAnswer(feed)
    feed = await hre.ethers.getContractAt(
      'AggregatorV3Interface',
      await assetContractLido.uoaPerTargetChainlinkFeed()
    )
    await updateAnswer(feed)
    feed = await hre.ethers.getContractAt(
      'AggregatorV3Interface',
      await assetContractLido.refPerTokenChainlinkFeed()
    )
    await updateAnswer(feed)
    feed = await hre.ethers.getContractAt(
      'AggregatorV3Interface',
      await assetContractLido.exchangeRateChainlinkFeed()
    )
    await updateAnswer(feed)
  } catch {
    // console.error('❌ targetPerRefChainlinkFeed, uoaPerTargetChainlinkFeed, or refPerTokenChainlinkFeed not found for:', asset, 'skipping...')
  }

  // targetPerTokChainlinkFeed
  try {
    const assetContractReth = await hre.ethers.getContractAt('RethCollateral', asset)
    const feed = await hre.ethers.getContractAt(
      'AggregatorV3Interface',
      await assetContractReth.targetPerTokChainlinkFeed()
    )
    await updateAnswer(feed)
  } catch {
    // console.error('❌ targetPerTokChainlinkFeed not found for:', asset, 'skipping...')
  }

  // Dealing with nested RTokens
  // TODO do better

  // eUSDFRAXBP
  if (
    asset == '0x890FAa00C16EAD6AA76F18A1A7fe9C40838F9122' ||
    asset == '0x5cD176b58a6FdBAa1aEFD0921935a730C62f03Ac' ||
    asset == '0x994455cE66Fd984e2A0A0aca453e637810a8f032'
  ) {
    const feed = await hre.ethers.getContractAt(
      'AggregatorV3Interface',
      networkConfig['1'].chainlinkFeeds.FRAX!
    )
    await updateAnswer(feed)
    const eUSDAssetRegistry = await hre.ethers.getContractAt(
      'IAssetRegistry',
      '0x9B85aC04A09c8C813c37de9B3d563C2D3F936162'
    )
    const [, eUSDAssets] = await eUSDAssetRegistry.getRegistry()
    for (const eUSDAsset of eUSDAssets) {
      addresses = await pushOracleForward(hre, eUSDAsset, addresses) // recursion!
    }
  }

  return addresses
}

export const setOraclePrice = async (
  hre: HardhatRuntimeEnvironment,
  asset: string,
  value: BigNumber
) => {
  const assetContract = await hre.ethers.getContractAt('TestIAsset', asset)
  const realChainlinkFeed = await hre.ethers.getContractAt(
    'AggregatorV3Interface',
    await assetContract.chainlinkFeed()
  )
  const oracle = await overrideOracle(hre, realChainlinkFeed.address)
  await oracle.updateAnswer(value)
}
