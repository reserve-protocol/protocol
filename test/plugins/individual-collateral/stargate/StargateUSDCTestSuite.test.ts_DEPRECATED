import collateralTests from '../collateralTests'
import { CollateralFixtureContext, CollateralOpts, MintCollateralFunc } from '../pluginTestTypes'
import { ethers } from 'hardhat'
import { ContractFactory, BigNumberish, BigNumber } from 'ethers'
import { resetFork } from './helpers'
import {
  ERC20Mock,
  MockV3Aggregator,
  MockV3Aggregator__factory,
  TestICollateral,
  StargatePoolMock,
  IStargateLPStaking,
  StargateRewardableWrapper,
  StargateRewardableWrapper__factory,
  IStargatePool,
} from '@typechain/index'
import { pushOracleForward } from '../../../utils/oracles'
import { bn, fp } from '#/common/numbers'
import { expect } from 'chai'
import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers'
import {
  STARGATE,
  USDC_USD_PRICE_FEED,
  MAX_TRADE_VOL,
  DEFAULT_THRESHOLD,
  DELAY_UNTIL_DEFAULT,
  SUSDC,
  ORACLE_TIMEOUT,
  ORACLE_ERROR,
  STAKING_CONTRACT,
  STARGATE_ROUTER,
  USDC_NAME,
} from './constants'
import { noop } from 'lodash'
import { whileImpersonating } from '#/test/utils/impersonation'
import { useEnv } from '#/utils/env'

/*
  Define interfaces
*/

interface StargateCollateralFixtureContext extends CollateralFixtureContext {
  pool: StargatePoolMock | IStargatePool
  wpool: StargateRewardableWrapper
  stargate: ERC20Mock
  stakingContract: IStargateLPStaking
}

export enum CollateralType {
  STABLE,
  VOLATILE,
}

export interface StargateCollateralOpts extends CollateralOpts {
  type?: CollateralType
}

/*
  Define deployment functions
*/

export const defaultStargateCollateralOpts: StargateCollateralOpts = {
  erc20: SUSDC,
  targetName: ethers.utils.formatBytes32String('USD'),
  rewardERC20: STARGATE,
  priceTimeout: ORACLE_TIMEOUT,
  chainlinkFeed: USDC_USD_PRICE_FEED,
  oracleTimeout: ORACLE_TIMEOUT,
  oracleError: ORACLE_ERROR,
  maxTradeVolume: MAX_TRADE_VOL,
  defaultThreshold: DEFAULT_THRESHOLD,
  delayUntilDefault: DELAY_UNTIL_DEFAULT,
  type: CollateralType.STABLE,
  revenueHiding: fp('0'),
}

export const deployCollateral = async (
  opts: StargateCollateralOpts = {}
): Promise<TestICollateral> => {
  opts = { ...defaultStargateCollateralOpts, ...opts }

  const StargatePoolCollateralFactory: ContractFactory = await ethers.getContractFactory(
    opts.type === CollateralType.STABLE ? 'StargatePoolFiatCollateral' : 'StargatePoolETHCollateral'
  )

  const collateral = <TestICollateral>await StargatePoolCollateralFactory.deploy(
    {
      erc20: opts.erc20,
      targetName: opts.targetName,
      priceTimeout: opts.priceTimeout,
      chainlinkFeed: opts.chainlinkFeed,
      oracleError: opts.oracleError,
      oracleTimeout: opts.oracleTimeout,
      maxTradeVolume: opts.maxTradeVolume,
      defaultThreshold: opts.defaultThreshold,
      delayUntilDefault: opts.delayUntilDefault,
    },
    opts.revenueHiding,
    { gasLimit: 2000000000 }
  )
  await collateral.deployed()

  // Push forward chainlink feed
  await pushOracleForward(opts.chainlinkFeed!)

  // sometimes we are trying to test a negative test case and we want this to fail silently
  // fortunately this syntax fails silently because our tools are terrible
  await expect(collateral.refresh())

  return collateral
}

type Fixture<T> = () => Promise<T>

const makeCollateralFixtureContext = (
  alice: SignerWithAddress,
  opts: StargateCollateralOpts = {}
): Fixture<StargateCollateralFixtureContext> => {
  const collateralOpts = { ...defaultStargateCollateralOpts, ...opts }

  const makeCollateralFixtureContext = async () => {
    return deployCollateralStargateMockContext(collateralOpts)
  }

  return makeCollateralFixtureContext
}

const deployCollateralStargateMockContext = async (
  opts: StargateCollateralOpts = {}
): Promise<StargateCollateralFixtureContext> => {
  const collateralOpts = { ...defaultStargateCollateralOpts, ...opts }

  const MockV3AggregatorFactory = <MockV3Aggregator__factory>(
    await ethers.getContractFactory('MockV3Aggregator')
  )
  let chainlinkFeed: MockV3Aggregator
  if (collateralOpts.type === CollateralType.STABLE)
    chainlinkFeed = <MockV3Aggregator>await MockV3AggregatorFactory.deploy(8, bn('1e8'))
  else {
    chainlinkFeed = <MockV3Aggregator>await MockV3AggregatorFactory.deploy(8, bn('1995e8'))
  }
  collateralOpts.chainlinkFeed = chainlinkFeed.address

  const StargateRewardableWrapperFactory = <StargateRewardableWrapper__factory>(
    await ethers.getContractFactory('StargateRewardableWrapper')
  )

  const stargate = await (
    await ethers.getContractFactory('ERC20Mock')
  ).deploy('Stargate Mocked token', 'S*MT')
  const stakingContract = await (
    await ethers.getContractFactory('StargateLPStakingMock')
  ).deploy(stargate.address)
  const mockPool = await (
    await ethers.getContractFactory('StargatePoolMock')
  ).deploy('Mock Pool', 'MSP', collateralOpts.type === CollateralType.STABLE ? 6 : 8)
  await stakingContract.add(bn('5000'), mockPool.address)
  await mockPool.mint(stakingContract.address, bn(1e6))
  await mockPool.setExchangeRate(fp(1))
  const pool = mockPool

  const wrapper = await StargateRewardableWrapperFactory.deploy(
    'Wrapped Stargate USDC TEST',
    'wsgUSDbC-TEST',
    stargate.address,
    stakingContract.address,
    pool.address
  )
  collateralOpts.erc20 = wrapper.address
  collateralOpts.rewardERC20 = stargate.address

  const collateral = await deployCollateral(collateralOpts)

  const rewardToken = <ERC20Mock>await ethers.getContractAt('ERC20Mock', STARGATE)

  return {
    collateral,
    chainlinkFeed,
    tok: wrapper,
    rewardToken,
    pool,
    wpool: wrapper,
    stargate,
    stakingContract,
  }
}

/*
  Define helper functions
*/

const mintCollateralTo: MintCollateralFunc<StargateCollateralFixtureContext> = async (
  ctx: StargateCollateralFixtureContext,
  amount: BigNumberish,
  user: SignerWithAddress,
  recipient: string
) => {
  const currentExchangeRate = await ctx.collateral.refPerTok()

  await whileImpersonating(STARGATE_ROUTER, async (router) => {
    await ctx.pool.connect(router).mint(user.address, amount)
  })
  await ctx.pool.connect(user).approve(ctx.wpool.address, ethers.constants.MaxUint256)

  await ctx.wpool.connect(user).deposit(amount, user.address)
  await ctx.wpool.connect(user).transfer(recipient, amount)
  if (ctx.pool.address != SUSDC) {
    ctx.pool.setExchangeRate(currentExchangeRate.add(fp('0.000001')))
  }
}

const reduceRefPerTok = async (
  ctx: StargateCollateralFixtureContext,
  pctDecrease: BigNumberish
) => {
  const currentExchangeRate = await ctx.pool.exchangeRate()
  await ctx.pool.setExchangeRate(
    currentExchangeRate.sub(currentExchangeRate.mul(pctDecrease).div(100))
  )
}

const increaseRefPerTok = async (
  ctx: StargateCollateralFixtureContext,
  pctIncrease: BigNumberish
) => {
  const currentExchangeRate = await ctx.pool.exchangeRate()
  await ctx.pool.setExchangeRate(
    currentExchangeRate.add(currentExchangeRate.mul(pctIncrease).div(100))
  )
}

const getExpectedPrice = async (ctx: StargateCollateralFixtureContext): Promise<BigNumber> => {
  const initRefPerTok = await ctx.collateral.refPerTok()

  const decimals = await ctx.chainlinkFeed.decimals()

  const initData = await ctx.chainlinkFeed.latestRoundData()
  return initData.answer
    .mul(bn(10).pow(18 - decimals))
    .mul(initRefPerTok)
    .div(fp('1'))
}

const reduceTargetPerRef = async (
  ctx: StargateCollateralFixtureContext,
  pctDecrease: BigNumberish
) => {
  const lastRound = await ctx.chainlinkFeed.latestRoundData()
  const nextAnswer = lastRound.answer.sub(lastRound.answer.mul(pctDecrease).div(100))
  await ctx.chainlinkFeed.updateAnswer(nextAnswer)
}

const increaseTargetPerRef = async (
  ctx: StargateCollateralFixtureContext,
  pctIncrease: BigNumberish
) => {
  const lastRound = await ctx.chainlinkFeed.latestRoundData()
  const nextAnswer = lastRound.answer.add(lastRound.answer.mul(pctIncrease).div(100))
  await ctx.chainlinkFeed.updateAnswer(nextAnswer)
}

const beforeEachRewardsTest = async (ctx: StargateCollateralFixtureContext) => {
  // switch to proper network rewards setup

  const stargate = <ERC20Mock>await ethers.getContractAt('ERC20Mock', STARGATE)
  const stakingContract = <IStargateLPStaking>(
    await ethers.getContractAt('IStargateLPStaking', STAKING_CONTRACT)
  )
  const pool: StargatePoolMock | IStargatePool = await ethers.getContractAt('IStargatePool', SUSDC)

  const StargateRewardableWrapperFactory = <StargateRewardableWrapper__factory>(
    await ethers.getContractFactory('StargateRewardableWrapper')
  )

  const wrapper = await StargateRewardableWrapperFactory.deploy(
    'Wrapped Stargate USDC TEST',
    'wsgUSDbC-TEST',
    stargate.address,
    stakingContract.address,
    pool.address
  )
  const opts = {
    erc20: wrapper.address,
    rewardERC20: stargate.address,
  }

  const collateral = await deployCollateral(opts)

  ctx.collateral = collateral
  ctx.pool = pool
  ctx.stakingContract = stakingContract
  ctx.stargate = stargate
  ctx.tok = wrapper
  ctx.wpool = wrapper
}

/*
  Run the test suite
*/

export const stableOpts = {
  deployCollateral,
  collateralSpecificConstructorTests: noop,
  collateralSpecificStatusTests: noop,
  beforeEachRewardsTest,
  makeCollateralFixtureContext,
  mintCollateralTo,
  reduceRefPerTok,
  increaseRefPerTok,
  resetFork,
  collateralName: `Stargate ${USDC_NAME} Pool`,
  reduceTargetPerRef,
  increaseTargetPerRef,
  itClaimsRewards: it, // reward growth not supported in mock
  itChecksTargetPerRefDefault: it,
  itChecksTargetPerRefDefaultUp: it,
  itChecksRefPerTokDefault: it,
  itChecksNonZeroDefaultThreshold: it,
  itHasRevenueHiding: it,
  itIsPricedByPeg: true,
  chainlinkDefaultAnswer: 1e8,
  itChecksPriceChanges: it,
  getExpectedPrice,
  targetNetwork: useEnv('FORK_NETWORK') ?? 'mainnet',
}

collateralTests(stableOpts)
