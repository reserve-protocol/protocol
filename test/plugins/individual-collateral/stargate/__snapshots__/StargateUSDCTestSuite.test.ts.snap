// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collateral: Stargate USDC Pool collateral functionality Gas Reporting ERC20 transfer 1`] = `109345`;

exports[`Collateral: Stargate USDC Pool collateral functionality Gas Reporting ERC20 transfer 2`] = `92245`;

exports[`Collateral: Stargate USDC Pool collateral functionality Gas Reporting refresh() after full price timeout 1`] = `56094`;

exports[`Collateral: Stargate USDC Pool collateral functionality Gas Reporting refresh() after full price timeout 2`] = `51626`;

exports[`Collateral: Stargate USDC Pool collateral functionality Gas Reporting refresh() after hard default 1`] = `74579`;

exports[`Collateral: Stargate USDC Pool collateral functionality Gas Reporting refresh() after hard default 2`] = `66897`;

exports[`Collateral: Stargate USDC Pool collateral functionality Gas Reporting refresh() after oracle timeout 1`] = `56094`;

exports[`Collateral: Stargate USDC Pool collateral functionality Gas Reporting refresh() after oracle timeout 2`] = `51626`;

exports[`Collateral: Stargate USDC Pool collateral functionality Gas Reporting refresh() after soft default 1`] = `66835`;

exports[`Collateral: Stargate USDC Pool collateral functionality Gas Reporting refresh() after soft default 2`] = `66835`;

exports[`Collateral: Stargate USDC Pool collateral functionality Gas Reporting refresh() during SOUND 1`] = `66894`;

exports[`Collateral: Stargate USDC Pool collateral functionality Gas Reporting refresh() during SOUND 2`] = `66894`;

exports[`Collateral: Stargate USDC Pool collateral functionality Gas Reporting refresh() during soft default 1`] = `74444`;

exports[`Collateral: Stargate USDC Pool collateral functionality Gas Reporting refresh() during soft default 2`] = `67176`;
